ReactiveCocoa/Swift/SignalProducer.swift

```diff
@@ -275,16 +285,29 @@ private struct BufferState<Value, Error: ErrorType> {
 	// producer was terminated.
 	var observers: Bag<Signal<Value, Error>.Observer>? = Bag()
 
-	// Appends a new value to the buffer, trimming it down to the given capacity
-	// if necessary.
-	mutating func addValue(value: Value, upToCapacity capacity: Int) {
-		values.append(value)
-
-		let overflow = values.count - capacity
-		if overflow > 0 {
-			values.removeRange(0..<overflow)
-		}
-	}
+    mutating func addValue(value: Value) {
+        if self.capacity == 1 {
+            self.values = [value]
+            return
+        }
+        if values.count == 0 {
+            values = Array<Value>(count: self.capacity * 2, repeatedValue: value)
+        }
+        values[index] = value
+        values[index + capacity] = value
+        index = (index + 1) % (self.capacity)
+        if count < self.capacity {
+            count = count + 1
+        }
+        self.slice = nil
+    }
+    
+    mutating func generator() -> IndexingGenerator<Array<Value>> {
```

Comment:
Could this be `generate()` and `BufferState` be `SequenceType`? Then the `forEach` loop in `replay` above could just be on `originalState` and avoid a force-unwrap.

(Code reviewing on my phone, so it's possible I’m missing something… pre-apologies if so)